### PR TITLE
DAOS-16307 client: Defer creating network context in child processes

### PR DESCRIPTION
**Pull request by wiliamhuang**
Features: ioil pil4dfs

Create network context when a read/write operation is processed.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by wiliamhuang** [comment]
All tests passed in CI.

File: src/client/dfuse/il/int_posix.c
Comment (mchaarawi) [inline src/client/dfuse/il/int_posix.c:809]: i don't think we should do this for IL.
IL only intercepts read/write calls unlike pil4dfs. So it's more likely that that context is going to be used in this case.

**Review by daltonbohning — APPROVED** [review]
LGTM but will defer to Mohamad.

FYI `Features: ioil` doesn't cover all libioil tests because the `il` tag is used too.
So you'd need `Features: ioil il`
I'll create a ticket to update those

File: src/client/dfuse/il/int_posix.c
Comment (wiliamhuang) [reply src/client/dfuse/il/int_posix.c:809]: child_hdlr() is executed once a child process is forked if the parent process calls open() to access files on daos. We do not know whether the child process only calls metadata related functions or not.

Comment (mchaarawi) [reply src/client/dfuse/il/int_posix.c:809]: this was not my point.
IOIL does not intercept any metadata or libaio calls, so creating a context here is not going to matter much and is only beneficial in case the child actually does IO that is intercepted.
But in pil4dfs case, the child will create additional network context now if it does libaio for example, and we already started creating aio context here taking away a resource from there.

Comment (wiliamhuang) [reply src/client/dfuse/il/int_posix.c]: ok. Just reverted the change in libioil.

**Review by daltonbohning — APPROVED** [review]

**Comment by wiliamhuang** [comment]
Two failed tests in CI.
1) intercept_multi_client.py:IorInterceptMultiClient.test_ior_intercept_libpil4dfs
This is just due to noisy performance data.
2) time out in pil4dfs_fio.py:Pil4dfsFio.test_pil4dfs_vs_dfs was reported in ticket https://daosio.atlassian.net/browse/DAOS-16315

**Comment by mchaarawi** [comment]
> Two failed tests in CI.
> 
> 1. intercept_multi_client.py:IorInterceptMultiClient.test_ior_intercept_libpil4dfs
>    This is just due to noisy performance data.
> 2. time out in pil4dfs_fio.py:Pil4dfsFio.test_pil4dfs_vs_dfs was reported in ticket https://daosio.atlassian.net/browse/DAOS-16315

TBH it's a bit risky to land changes that actually do effect existing tests that are intermittently failing. let's hold this and investigate the other issues first before landing this one.

**Comment by wiliamhuang** [comment]
> > Two failed tests in CI.
> > 
> > 1. intercept_multi_client.py:IorInterceptMultiClient.test_ior_intercept_libpil4dfs
> >    This is just due to noisy performance data.
> > 2. time out in pil4dfs_fio.py:Pil4dfsFio.test_pil4dfs_vs_dfs was reported in ticket https://daosio.atlassian.net/browse/DAOS-16315
> 
> TBH it's a bit risky to land changes that actually do effect existing tests that are intermittently failing. let's hold this and investigate the other issues first before landing this one.

ok. I am working on https://daosio.atlassian.net/browse/DAOS-16315. Thank you!

**Comment by wiliamhuang** [comment]
> > Two failed tests in CI.
> > 
> > 1. intercept_multi_client.py:IorInterceptMultiClient.test_ior_intercept_libpil4dfs
> >    This is just due to noisy performance data.
> > 2. time out in pil4dfs_fio.py:Pil4dfsFio.test_pil4dfs_vs_dfs was reported in ticket https://daosio.atlassian.net/browse/DAOS-16315
> 
> TBH it's a bit risky to land changes that actually do effect existing tests that are intermittently failing. let's hold this and investigate the other issues first before landing this one.

Can we land this PR now? The failure of libaio + fork() in fio is caused by mercury 2.4.0rc4.

**Pull request merged**
